src/mame/misc/astrcorp.cpp

```diff
@@ -1,58 +1,168 @@
+- showhand: after pressing payout (I), press start while "COLLECTING POINTS" is displayed to activate the hopper.
+  Otherwise, while "GAME" is displayed, the points just vanish.
+- showhand: hopper jam error if "payout sensor type" is set to "normal low" in the settings.
+  It seems the game code is buggy: it uses 4000 as hopper status mask (from showhanc) instead of 8000.
+- winbingo: it starts with a Tetris game. To switch to the gambling game, press stop all + stop 2 (Z + C).
 
 TODO:
 
-- Find source of level 2 interrupt (sprite DMA end?);
-- magibomb: fix timings;
-- astoneag, dinodino, magibombd: currently stop with 'ROM error'. The first few hundred bytes don't decrypt correctly,
-  it is suspected there's a ROM overlay provided by the protection device;
-- winbingo and clones, zulu: these should be at the same state of the two above, but don't show anything on screen;
-- magibomba, westvent: need a redump of one of the program ROMs;
+- MCU on showhand/showhanc PCB, what does it do? Games run fine and there appears to be no interaction...
+- Protection in magibomb sets with simpler encryption (writes to $a0101x, $a11100, etc.). Some sets appear to work fine regardless, others freeze.
+- Fix mirror ROM checksum / ROM overlay without code patches (in games with heavier encryption).
+- Find source of level 2 interrupt (sprite DMA end?).
+- magibomba, westvent: need a redump of one of the program ROMs.
+- magibombd, hacher: need a redump of the sprite ROMs.
+- astoneag, dinodino, magibombd: exiting from test menu goes haywire (requires a soft-reset with F3).
 
 *************************************************************************************************************/
 
 #include "emu.h"
+
 #include "cpu/m68000/m68000.h"
 #include "machine/eepromser.h"
 #include "machine/nvram.h"
 #include "machine/ticket.h"
 #include "machine/timer.h"
 #include "sound/okim6295.h"
 #include "video/ramdac.h"
+
 #include "emupal.h"
 #include "screen.h"
 #include "speaker.h"
 
+#include "astoneag.lh"
+#include "dinodino.lh"
+#include "magibomb.lh"
+#include "magibombb.lh"
+#include "showhanc.lh"
+#include "showhand.lh"
+#include "skilldrp.lh"
+#include "winbingo.lh"
+#include "zoo.lh"
+
+/***************************************************************************
+
+    CPU Code (Protection)
+
+***************************************************************************/
+
+class astro_cpucode_device : public device_t
+{
+public:
+	astro_cpucode_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock);
+
+	// read handlers
+	DECLARE_READ_LINE_MEMBER(do_read);    // DO
+
+	// write handlers
+	DECLARE_WRITE_LINE_MEMBER(cs_write);  // CS signal (active high)
+	DECLARE_WRITE_LINE_MEMBER(clk_write); // CLK signal (active high)
+
+protected:
+	virtual void device_start() override;
+	virtual void device_reset() override;
+
+private:
+	required_memory_region m_region_key;
```

Comment:
If you use `required_region_ptr<T>` for this, you can index it directly without the calls to `->base()` – it makes the code cleaner.

Reply:
Done.